drivers/i2s/i2s_mcux_sai.c

```diff
@@ -0,0 +1,1172 @@
+
+	/* Disable the channel FIFO */
+	dev_cfg->base->RCR3 &= ~I2S_RCR3_RCE_MASK;
+
+	/* Disable DMA enable bit */
+	SAI_RxEnableDMA(dev_cfg->base, kSAI_FIFORequestDMAEnable,
+			false);
+
+	/* Disable Rx */
+	SAI_RxEnable(dev_cfg->base, false);
+
+	/* purge buffers queued in the stream */
+	i2s_purge_stream_buffers(strm, dev_data->rx.cfg.mem_slab);
+
+	strm->stream_starving = false;
+
+	/* If Rx is disabled, reset the FIFO pointer
+	 * and clear error flags
+	 */
+	if ((dev_cfg->base->RCSR & I2S_RCSR_RE_MASK) == 0UL) {
+		dev_cfg->base->RCSR |=
+			(I2S_RCSR_FR_MASK | I2S_RCSR_SR_MASK);
+		dev_cfg->base->RCSR &= ~I2S_RCSR_SR_MASK;
+	}
+}
+
+/* This function is executed in the interrupt context */
+static void i2s_dma_tx_callback(const struct device *dma_dev,
+				void *arg, uint32_t channel, int status)
+{
+	const struct device *dev = (struct device *)arg;
+	const struct i2s_mcux_config *dev_cfg =
+		(const struct i2s_mcux_config *)dev->config;
+	I2S_Type *base = (I2S_Type *)dev_cfg->base;
+	struct i2s_dev_data *dev_data = (struct i2s_dev_data *)dev->data;
+	struct stream *strm = &dev_data->tx;
+	void *buffer = NULL;
+	int ret;
+
+	LOG_DBG("tx cb");
+
+	ret = k_msgq_get(&strm->out_queue, &buffer, K_NO_WAIT);
+	if (ret == 0) {
+		/* transmission complete. free the buffer */
+		k_mem_slab_free(strm->cfg.mem_slab, &buffer);
+	} else {
+		LOG_ERR("no buf in out queue for channel %u", channel);
+	}
+
+	switch (strm->state) {
+	case I2S_STATE_RUNNING:
+		/* get the next buffer from queue */
+		if (strm->in_queue.used_msgs == 0) {
+			LOG_DBG("tx no more in queue data");
+			i2s_tx_stream_disable(dev);
+			/* stream is starving */
+			strm->stream_starving = true;
+			return;
+		}
+		ret = k_msgq_get(&strm->in_queue, &buffer, K_NO_WAIT);
+		if (ret == 0) {
+			/* config the DMA */
+			uint32_t data_path = dev_data->tx.start_channel;
+
+			strm->dma_block.block_size = strm->cfg.block_size;
+			strm->dma_block.dest_address =
+				(uint32_t)&base->TDR[data_path];
+			strm->dma_block.source_address = (uint32_t)buffer;
+			dma_config(dev_data->dev_dma, strm->dma_channel,
+				   &strm->dma_cfg);
+			ret = k_msgq_put(&strm->out_queue,
+					 &buffer, K_NO_WAIT);
+			if (ret != 0) {
+				LOG_ERR("buffer %p -> out %p err %d",
+					buffer, &strm->out_queue, ret);
+			}
+			dma_start(dev_data->dev_dma, strm->dma_channel);
+		}
+
+		if (ret || status) {
+		/*
+		 * DMA encountered an error (status != 0)
+		 * or No bufers in input queue
+		 */
+			LOG_ERR("DMA status %08x chan %u get ret %d",
+				status, channel, ret);
+			strm->state = I2S_STATE_READY;
+			i2s_tx_stream_disable(dev);
+		}
+
+		break;
+
+	case I2S_STATE_STOPPING:
+		i2s_tx_stream_disable(dev);
+		strm->state = I2S_STATE_READY;
+		break;
+	}
+}
+
+static void i2s_dma_rx_callback(const struct device *dma_dev,
+				void *arg, uint32_t channel, int status)
+{
+	struct device *dev = (struct device *)arg;
+	const struct i2s_mcux_config *dev_cfg =
+		(const struct i2s_mcux_config *)dev->config;
```

Comment:
I am not asking for a change, but casting a void* to something else is not necessary in C. You can just write:

```
struct device *dev = arg;
const struct i2s_mcux_config *dev_cfg = dev->config;
```

That is why I mentioned removing the extra macros in this comment: https://github.com/zephyrproject-rtos/zephyr/pull/37335#discussion_r727367542

They are not needed to save typing.

Reply:
right, let me update.